### ISSUE TITLE
A weird performance of the new release!

### ISSUE THREAD

**Issue by Jameel9** [chemprop/chemprop#1168]
Hi @JacksonBurns and @KnathanM, 

Sorry for being the annoying user 😆, but I do think that there is a problem in the overall performance of the new release!
I have a dataset (8,768 mols) that I have split externally (and randomly) into 90% (train_valid_subset.csv) and 10% (external_test_subset.csv) into separated csv files to ensure consistency and reduce the chances of human error. The 10% file was kept for final testing and plotting. When I used the 90% file for building and optimizing a binary classification model, the performance of v2.1.1 seems to be significantly lower than v2.0.1 (installed on two machines, including the machine containing the final release). That sounds unreal but it did happen. I used the same command on the same computer (office PC) to compare (i) the latest release, (ii) the recently installed 2.0.4 version (both on the office PC) and (iii) the installation on my laptop that happened in September 2024. For consistency, the following command was used across all the runs

# For the 2.0.4 
`chemprop hpopt --search-parameter-keywords all --data-path train_valid_subsets.csv --task-type classification --smiles-columns smiles --target-columns label --epochs 100 --num-folds 3 --ensemble-size 5 --metrics {'roc','binary-mcc','prc','accuracy','f1'} --num-workers 20 --hpopt-save-dir optimization_clss --raytune-num-workers 1 --raytune-use-gpu`

# For the 2.1.1
`chemprop hpopt --search-parameter-keywords all --data-path train_valid_subsets.csv --task-type classification --smiles-columns smiles --target-columns label --epochs 100 --num-replicates 3 --ensemble-size 5 --metrics {'roc','binary-mcc','prc','accuracy','f1'} --num-workers 20 --hpopt-save-dir optimization_clss --raytune-num-workers 1 --raytune-use-gpu`

The different between final5 and final3 is that for the number of num-replicates during training/optimization (one run was set to 3 and in the other to 5), and both runs were on the same machine. After the hpopt finish, I used the "best_checkpoint.ckpt" file to predict the holdout 10% and did some plots. The attached plot is for the attached dataset. It would be great if you replicate this scenario in your machine(s) and see if the problem is in my end. 

![Image](https://github.com/user-attachments/assets/57371961-5247-4107-ac7a-f0fe59fcec58)

recovered20 means the recently installed version 2.0.4. The dataset is attached as a zip file
final is the final release (2.1.1)
[for_jackson.zip](https://github.com/user-attachments/files/18859926/for_jackson.zip)

**Comment by JacksonBurns** [comment]
When you specify `--num-folds 3` in version 2.0.4 you end up with a 33/33/33 (train/val/test) split, whereas specifying `--num-replicates 3` (or 5) will use the default split size of 80/10/10 each time. Looks like your model is overfitting with this small validation size. Consider changing your split sizes in 2.1 to look more like 2.0 and see how things look.

Also bear in mind that the data reserved for testing will be thrown out (just not used) during `hpopt`

**Comment by JacksonBurns** [comment]
@KnathanM please check my statement above - memory is already foggy about how `hpopt` would interact with `--num-folds` back in 2.0

**Comment by Jameel9** [comment]
@JacksonBurns 
Is n't it that `--num-folds 3` is the folds for cross validation and the split is always 80/10/10?! while` --num-replicates 3 ` is the number of models trained on each split?

**Comment by Jameel9** [comment]
@JacksonBurns 
So, if I specified `--split-sizes 0.33 0.33 0.33` in 2.1.1 and kept everything else in the command line (above) as it is, it should replicate the results seen from the 2.0.4?

**Comment by KnathanM** [comment]
Our documentation for <2.1 was incorrect about `--num-folds`. It said 
> k-fold cross-validation can be run by specifying `--num-folds <k>` (default 1, i.e. no cross-validation)
which makes it sounds like specifying `--num-folds` was enough. But actually that flag does nothing if you don't specify --split cv. So the 2.0.4 command should only train 5 models instead of 15, all trained on a single 80/10/10 random split. In any event, the hpopt only uses the first split to find optimal hyperparameters (we'd like to change this in the future) so the hpopt should be the same between 2.0.4 and 2.1.1. Are the hyperparameters for the 2.0.4 models similar to the hyperparameters for the 2.1.1? I guess they don't have to be because the hpopt process depends on the exact split and I don't think the search is deterministic either. 

I ran this command using the data you provided in 2.1.1.
```
chemprop hpopt \
--search-parameter-keywords all \
--data-path train_valid_subsets.csv \
--task-type classification \
--smiles-columns smiles \
--target-columns label \
--epochs 100 \
--num-replicates 3 \
--ensemble-size 5 \
--metrics {'roc','binary-mcc','prc','accuracy','f1'} \
--num-workers 20 \
--hpopt-save-dir optimization_clss \
--raytune-temp-dir $RAY_TEMP_DIR \
--raytune-num-cpus 40 \
--raytune-num-gpus 2 \
--raytune-max-concurrent-trials 2 \
--hyperopt-random-state-seed 42 \

#Training with optimized hyperparameters
chemprop train \
-t classification \
--data-path train_valid_subsets.csv \
--num-workers 20 \
--epochs 100 \
--pytorch-seed 42 \
--save-dir optimization_clss \
--ensemble-size 5 \
--metrics {'roc','binary-mcc','prc','accuracy','f1'} \
--config-path optimization_clss/best_config.toml \
--num-replicates 3 \
--smiles-columns smiles \
--target-columns label \
```

And got this result
![Image](https://github.com/user-attachments/assets/c7f284ab-5cb0-4d29-935e-92fa03907918)

So it looks like 2.1.1. works okay. I'm not sure what happened to cause the poor performance on your end. The best_config.toml is
```
smiles-columns = [smiles]
num-workers = 20
data-path = train_valid_subsets.csv
ensemble-size = 5
target-columns = [label]
task-type = classification
metrics = [roc, binary-mcc, prc, accuracy, f1]
epochs = 100
num-replicates = 3
batch-size = 16
accelerator = auto
devices = auto
rxn-mode = REAC_DIFF
multi-hot-atom-featurizer-mode = V2
frzn-ffn-layers = 0
message-hidden-dim = 2300
depth = 4
dropout = 0.2
activation = LEAKYRELU
aggregation = norm
aggregation-norm = 146.0
ffn-hidden-dim = 1700
ffn-num-layers = 1
multiclass-num-classes = 3
v-kl = 0.0
eps = 1e-08
alpha = 0.1
tracking-metric = val_loss
warmup-epochs = 6
init-lr = 5.933949556291671e-05
max-lr = 0.0038657504949304597
final-lr = 0.0010384743843879292
split = RANDOM
split-sizes = [0.8, 0.1, 0.1]
split-key-molecule = 0
data-seed = 0
```

**Comment by KnathanM** [comment]
It looks like there isn't a problem with 2.1.1 based on my test, but if you have further questions/discussion on this topic, feel free to reopen the issue.
